test/reload_evolution/storage.test.lua

```diff
@@ -23,13 +23,17 @@ test_run:switch('storage_1_a')
 vshard.storage.bucket_force_create(1, vshard.consts.DEFAULT_BUCKET_COUNT / 2)
 bucket_id_to_move = vshard.consts.DEFAULT_BUCKET_COUNT
 
+test_run:switch('storage_2_b')
```

Comment:
Ah, don't like that. Can't really get, what is going on inside this test. Without turning the protection off on the `storage_2_b` I got the following log:

<details>
<summary>storage_2_b.log</summary>

```text
Starting instance storage_2_b...
started
Starting configuration of replica 001688c3-66f8-4a31-8e19-036c17d489c2
Run console at unix/:/home/alver/Programming/Work/vshard/test/var/001_reload_evolution/storage_2_b.control
started
mkdir /home/alver/Programming/Work/vshard/test/var/001_reload_evolution/storage_2_b
2022-11-13 18:27:53.971 [105742] main/103/storage_2_b I> Tarantool 2.11.0-entrypoint-697-gbab289cd8
2022-11-13 18:27:53.971 [105742] main/103/storage_2_b I> log level 5
2022-11-13 18:27:53.971 [105742] main/103/storage_2_b I> wal/engine cleanup is paused
2022-11-13 18:27:53.972 [105742] main/103/storage_2_b I> mapping 268435456 bytes for memtx tuple arena...
2022-11-13 18:27:53.972 [105742] main/103/storage_2_b I> Actual slab_alloc_factor calculated on the basis of desired slab_alloc_factor = 1.044274
2022-11-13 18:27:53.972 [105742] main/103/storage_2_b I> mapping 134217728 bytes for vinyl tuple arena...
2022-11-13 18:27:53.975 [105742] main/103/storage_2_b I> update replication_synchro_quorum = 1
2022-11-13 18:27:53.975 [105742] main/103/storage_2_b I> instance uuid 001688c3-66f8-4a31-8e19-036c17d489c2
2022-11-13 18:27:53.975 [105742] main/103/storage_2_b I> tx_binary: bound to 127.0.0.1:3304
2022-11-13 18:27:53.975 [105742] main/103/storage_2_b I> connecting to 2 replicas
2022-11-13 18:27:53.976 [105742] main/112/applier/storage@127.0.0.1:3303 I> remote master 1e02ae8a-afc0-4e91-ba34-843a356b8ed7 at 127.0.0.1:3303 running Tarantool 2.11.0
2022-11-13 18:27:53.976 [105742] main/113/applier/storage@127.0.0.1:3304 I> remote master 001688c3-66f8-4a31-8e19-036c17d489c2 at 127.0.0.1:3304 running Tarantool 2.11.0
2022-11-13 18:27:53.976 [105742] main/103/storage_2_b I> connected to 2 replicas
2022-11-13 18:27:53.976 [105742] main/112/applier/storage@127.0.0.1:3303 I> failed to authenticate
2022-11-13 18:27:53.976 [105742] main/112/applier/storage@127.0.0.1:3303 box.cc:3183 E> ER_LOADING: Instance bootstrap hasn't finished yet
2022-11-13 18:27:53.976 [105742] main/112/applier/storage@127.0.0.1:3303 I> will retry every 0.10 second
2022-11-13 18:27:54.079 [105742] main/112/applier/storage@127.0.0.1:3303 I> authenticated
2022-11-13 18:27:54.079 [105742] main/103/storage_2_b I> bootstrapping replica from 1e02ae8a-afc0-4e91-ba34-843a356b8ed7 at 127.0.0.1:3303
2022-11-13 18:27:54.082 [105742] main/112/applier/storage@127.0.0.1:3303 I> cluster uuid ac522f65-aa94-4134-9f64-51ee384f1a54
2022-11-13 18:27:54.088 [105742] main/112/applier/storage@127.0.0.1:3303 I> assigned id 1 to replica 1e02ae8a-afc0-4e91-ba34-843a356b8ed7
2022-11-13 18:27:54.088 [105742] main/112/applier/storage@127.0.0.1:3303 I> update replication_synchro_quorum = 1
2022-11-13 18:27:54.088 [105742] main/112/applier/storage@127.0.0.1:3303 I> initial data received
2022-11-13 18:27:54.088 [105742] main/112/applier/storage@127.0.0.1:3303 I> assigned id 2 to replica 001688c3-66f8-4a31-8e19-036c17d489c2
2022-11-13 18:27:54.088 [105742] main/112/applier/storage@127.0.0.1:3303 I> update replication_synchro_quorum = 2
2022-11-13 18:27:54.088 [105742] main/112/applier/storage@127.0.0.1:3303 I> RAFT: fencing paused
2022-11-13 18:27:54.088 [105742] main/112/applier/storage@127.0.0.1:3303 I> final data received
2022-11-13 18:27:54.090 [105742] snapshot/101/main I> saving snapshot `/home/alver/Programming/Work/vshard/test/var/001_reload_evolution/storage_2_b/00000000000000000038.snap.inprogress'
2022-11-13 18:27:54.091 [105742] snapshot/101/main I> done
2022-11-13 18:27:54.091 [105742] main/103/storage_2_b I> ready to accept requests
2022-11-13 18:27:54.091 [105742] main/103/storage_2_b I> leaving orphan mode
2022-11-13 18:27:54.091 [105742] main/113/applier/storage@127.0.0.1:3304 I> leaving orphan mode
2022-11-13 18:27:54.091 [105742] main/103/storage_2_b I> set 'replication_connect_timeout' configuration option to 3
2022-11-13 18:27:54.091 [105742] main/103/storage_2_b I> set 'replication_timeout' configuration option to 0.1
2022-11-13 18:27:54.091 [105742] main/103/storage_2_b I> set 'log' configuration option to "\/home\/alver\/Programming\/Work\/vshard\/test\/var\/001_reload_evolution\/storage_2_b.log"
2022-11-13 18:27:54.092 [105742] main/103/storage_2_b I> set 'replication' configuration option to ["storage@127.0.0.1:3303","storage@127.0.0.1:3304"]
2022-11-13 18:27:54.092 [105742] main/103/storage_2_b I> set 'listen' configuration option to "storage:storage@127.0.0.1:3304"
2022-11-13 18:27:54.092 [105742] main/103/storage_2_b I> set 'instance_uuid' configuration option to "001688c3-66f8-4a31-8e19-036c17d489c2"
2022-11-13 18:27:54.092 [105742] main/103/storage_2_b I> set 'replication_connect_quorum' configuration option to 0
2022-11-13 18:27:54.092 [105742] main/103/storage_2_b I> set 'replicaset_uuid' configuration option to "ac522f65-aa94-4134-9f64-51ee384f1a54"
2022-11-13 18:27:54.092 [105742] main/107/checkpoint_daemon I> scheduled next checkpoint for Sun Nov 13 19:48:24 2022
2022-11-13 18:27:54.092 [105742] main/112/applier/storage@127.0.0.1:3303 I> subscribed
2022-11-13 18:27:54.092 [105742] main/112/applier/storage@127.0.0.1:3303 I> remote vclock {1: 38} local vclock {1: 38}
2022-11-13 18:27:54.092 [105742] main/103/storage_2_b I> set 'read_only' configuration option to true
2022-11-13 18:27:54.092 [105742] main/103/storage_2_b I> Box has been configured
2022-11-13 18:27:54.092 [105742] main/103/storage_2_b I> Replica' vshard schema version is not latest - current {0.1.15.0} vs latest {0.1.16.0}, but the replica still can work
2022-11-13 18:27:54.094 [105742] main I> entering the event loop
2022-11-13 18:27:54.094 [105742] main/112/applier/storage@127.0.0.1:3303 I> RAFT: message {term: 1, state: follower} from 1
2022-11-13 18:27:54.094 [105742] main/112/applier/storage@127.0.0.1:3303 I> leaving orphan mode
2022-11-13 18:27:54.174 [105742] main/111/main I> subscribed replica 1e02ae8a-afc0-4e91-ba34-843a356b8ed7 at fd 26, aka 127.0.0.1:3304, peer of 127.0.0.1:54998
2022-11-13 18:27:54.174 [105742] main/111/main I> remote vclock {1: 67} local vclock {1: 67}
2022-11-13 18:27:54.175 [105742] main/106/gc I> wal/engine cleanup is resumed
2022-11-13 18:27:54.175 [105742] relay/127.0.0.1:54998/101/main I> recover from `/home/alver/Programming/Work/vshard/test/var/001_reload_evolution/storage_2_b/00000000000000000038.xlog'
2022-11-13 18:27:54.189 [105742] main/112/applier/storage@127.0.0.1:3303 applier.cc:311 E> error applying row: {type: 'INSERT', replica_id: 1, lsn: 68, space_id: 512, index_id: 0, tuple: [1501, "active"]}
2022-11-13 18:27:54.189 [105742] main/112/applier/storage@127.0.0.1:3303 I> can't read row
2022-11-13 18:27:54.189 [105742] main/112/applier/storage@127.0.0.1:3303 utils.c:489 E> LuajitError: {"bucket_id":1501,"reason":"transition 'cdata<void *>: NULL' to 'active' is not allowed","code":39,"type":"ShardingError","message":"Bucket 1501 update is invalid: transition 'cdata<void *>: NULL' to 'active' is not allowed","name":"BUCKET_INVALID_UPDATE"}
2022-11-13 18:28:15.131 [105742] main C> got signal 2 - Interrupt
2022-11-13 18:28:15.132 [105742] main/129/iproto.shutdown I> tx_binary: stopped
```

</details>

What is going on? AFAICS, we don't reload `storage_2_b`. Why does it cannot recover from xlog? And moreover why does it at all trying to recover from it? We don't reconfigure it at all. Help would be really appreciated

Reply:
All is fine here. All storages are started in the new version except `storage_2_a`, which is started from an old vshard and then reloaded. Hence you have the protection enabled on `storage_2_b.`

Reply:
Got you, thanks